### ISSUE TITLE
ICU select syntax fails to compile when cases are numbers

### ISSUE THREAD

**Issue by Noah-Houghton** [flutter/flutter#120712]
I wasn't sure whether to put this here or over at https://github.com/dart-lang/intl -- happy to move it there if it'd be more appropriate.

## Steps to Reproduce
1. Create localization ARB file where an item contains a select statement utilizing numbers (see below for example)
2. Run `flutter gen-l10n`
3. Observe syntax parsing error

**Expected results:**
Parsing succeeds and the ARB file generates logic to provide the correct translated string on the provided values.

**Actual results:**
Parsing fails with the following error:

```
[localization_en.arb:dateSuffix] ICU Syntax Error: Expected "}" but found "1".
    {date, select, 1{st} 2{nd} 3{rd} 21{st} 22{nd} 23{rd} 31{st} other{th}}
                   ^
```

<details>
<summary>Code sample</summary>
I can upload the full recreation project if you like, but it's easy enough to create by adding the following pieces to any existing project:

```json
// lib/l10n/localization_en.arb
{
  "dateSuffix": "{date, select, 1{st} 2{nd} 3{rd} 21{st} 22{nd} 23{rd} 31{st} other{th}}",
  "@dateSuffix": {
    "type": "text",
    "description": "Ordinal suffix for dates",
    "placeholders": {
      "date": {
        "type": "int"
      }
    }
  }
}
```

```yaml
# l10n.yaml
arb-dir: lib/l10n
template-arb-file: localization_en.arb
output-localization-file: app_localizations.dart
untranslated-messages-file: untranslated_messages.txt
nullable-getter: false
```

</details>

<details>
  <summary>Logs</summary>

```
[localization_en.arb:dateSuffix] ICU Syntax Error: Expected "}" but found "1".
    {date, select, 1{st} 2{nd} 3{rd} 21{st} 22{nd} 23{rd} 31{st} other{th}}
                   ^
Running "flutter pub get" in localization_test...
Resolving dependencies... 
  collection 1.17.0 (1.17.1 available)
  js 0.6.5 (0.6.7 available)
  matcher 0.12.13 (0.12.14 available)
  meta 1.8.0 (1.9.0 available)
  path 1.8.2 (1.8.3 available)
  test_api 0.4.16 (0.4.18 available)
Got dependencies!
Analyzing localization_test...                                          
No issues found! (ran in 1.0s)

```

```
[✓] Flutter (Channel stable, 3.7.3, on Ubuntu 22.04.1 LTS 5.15.0-60-generic, locale en_US.UTF-8)
    • Flutter version 3.7.3 on channel stable at /home/noah/snap/flutter/common/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 9944297138 (6 days ago), 2023-02-08 15:46:04 -0800
    • Engine revision 248290d6d5
    • Dart version 2.19.2
    • DevTools version 2.20.1

[✓] Android toolchain - develop for Android devices (Android SDK version 32.0.0)
    • Android SDK at /home/noah/Android/Sdk
    • Platform android-33, build-tools 32.0.0
    • Java binary at: /home/noah/dev/apps/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.15+0-b2043.56-8887301)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop
    • clang version 10.0.0-4ubuntu1
    • cmake version 3.16.3
    • ninja version 1.10.0
    • pkg-config version 0.29.1

[✓] Android Studio (version 2022.1)
    • Android Studio at /home/noah/dev/apps/android-studio
    • Flutter plugin version 72.0.2
    • Dart plugin version 221.6096
    • Java version OpenJDK Runtime Environment (build 11.0.15+0-b2043.56-8887301)

[✓] VS Code
    • VS Code at /snap/code/current
    • Flutter extension version 3.58.0

[✓] Connected device (2 available)
    • Linux (desktop) • linux  • linux-x64      • Ubuntu 22.04.1 LTS 5.15.0-60-generic
    • Chrome (web)    • chrome • web-javascript • Google Chrome 110.0.5481.77

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!

```

</details>

**Comment by darshankawar** [comment]
Thanks for the report @Noah-Houghton 

> I wasn't sure whether to put this here or over at https://github.com/dart-lang/intl -- happy to move it there if it'd be more appropriate.

This is the correct place to report such errors.

Regarding the error you reported, I do see the same upon running it on latest stable, as below:

```
dhs@Dhss-MacBook-Pro test_app % flutter gen-l10n
Because l10n.yaml exists, the options defined there will be used instead.
To use the command line arguments, delete the l10n.yaml file in the Flutter project.


[app_en.arb:dateSuffix] ICU Syntax Error: Expected "}" but found "1".
    {date, select, 1{st} 2{nd} 3{rd} 21{st} 22{nd} 23{rd} 31{st} other{th}}
                   ^
dhs@Dhss-MacBook-Pro test_app % 
```

Although switching to latest master and running it as-is first gives me below log:

```
dhs@Dhss-MacBook-Pro test_app % flutter gen-l10n
Because l10n.yaml exists, the options defined there will be used instead.
To use the command line arguments, delete the l10n.yaml file in the Flutter project.


Placeholders used in selects must be of type 'String'
```



> ```json
> "date": {
>         "type": "int"
>       }
> ```

It is complaining about the `type` you passed which is expected to have `String`. So after updating the type and running `flutter gen-l10n` doesn't give me the reported error, but below as expected:


```
dhs@Dhss-MacBook-Pro test_app % flutter gen-l10n
Because l10n.yaml exists, the options defined there will be used instead.
To use the command line arguments, delete the l10n.yaml file in the Flutter project.


dhs@Dhss-MacBook-Pro test_app % 

```

I also see the `app_localizations_en.dart` generating properly after above step and it shows me below:

```
import 'package:intl/intl.dart' as intl;

import 'app_localizations.dart';

/// The translations for English (`en`).
class AppLocalizationsEn extends AppLocalizations {
  AppLocalizationsEn([String locale = 'en']) : super(locale);

  @override
  String dateSuffix(String date) {
    String _temp0 = intl.Intl.selectLogic(
      date,
      {
        '1': 'st',
        '2': 'nd',
        '3': 'rd',
        '21': 'st',
        '22': 'nd',
        '23': 'rd',
        '31': 'st',
        'other': 'th',
      },
    );
    return '$_temp0';
  }
}
```

In short, the error you reported doesn't occur on latest master version so you may want to switch to it until the fix propogates to stable in next release.

Also, check this new addition recently introduced for your further reference: https://docs.flutter.dev/development/accessibility-and-localization/internationalization#escaping-syntax

Although I tried to use that flag but it didn't help but just for your reference.

**Comment by Noah-Houghton** [comment]
@darshankawar Confirmed your solution works -- thank you for the quick response!
